### PR TITLE
primitives: add openCL support and autodection of best implementation

### PR DESCRIPTION
**Pull request by hardening**
This PR adds support for OpenCL in primitives, for now only YUVtoRGB primitives are implemented.

### CODE REVIEW

**Comment by akallabeth** [comment]
@hardening did some cleanups here: https://github.com/akallabeth/FreeRDP/tree/opencl
One thing that may improve speed a lot is implementing `YUV444ToRGB_8u_P3AC4R` as `YUV420ToRGB_8u_P3AC4R` is currently only used by the video redirection (`AVC420` is more or less dead)

File: libfreerdp/primitives/primitives.c [hunk 1]
```diff
@@ -23,65 +24,483 @@
...
+
+	if (!gotGPU)
+	{
+		WLog_ERR(TAG, "openCL: no GPU found");
+		return FALSE;
+	}
+
+	snprintf(sourcePath, sizeof(sourcePath), "%s/primitives.cl", OPENCL_SOURCE_PATH);
+
+	f = fopen(sourcePath, "r");
```
Comment (akallabeth) [inline libfreerdp/primitives/primitives.c]: Why import that at runtime? While nice to allow changing the implementation it is a hindrance if you compile a single binary (like `wfreerdp.exe`)

Comment (hardening) [reply libfreerdp/primitives/primitives.c]: well to my knowledge there's no way to have a precompiled version as it depends on the driver, it's like a shader. The other solution would be to embed the source in the freerdp binary...

Comment (akallabeth) [reply libfreerdp/primitives/primitives.c]: this could, of course, be done optionally ;)

**Comment by akallabeth** [comment]
@hardening any update on this? would love to merge ;)

**Comment by hardening** [comment]
> `FindOpenCL.cmake` is shipped with `CMake` 3.2 and above.
> The variables differ to your script though, so maybe it would be good to fix that.
> https://cmake.org/cmake/help/latest/module/FindOpenCL.html

Unfortunately for now the minimum cmake version that we require is 2.8. Do you suggest changing my file so that it match the shipped script ?

**Comment by akallabeth** [comment]
@hardening would use the variables of the module available since `3.2` and add a check to use the cmake provided version for these cases. (maybe like the `cmake/compat*` folders?)

File: libfreerdp/primitives/primitives.c [hunk 2]
```diff
@@ -23,72 +24,377 @@
...
 }
+
+primitives_t* primitives_get_by_type(DWORD type)
+{
+	InitOnceExecuteOnce(&generic_primitives_InitOnce, primitives_init_generic_cb, NULL, NULL);
+	switch (type)
+	{
+#if defined(WITH_OPENCL)
+		case PRIMITIVES_ONLY_GPU:
+			InitOnceExecuteOnce(&gpu_primitives_InitOnce, primitives_init_cpu_cb, NULL, NULL);
```
Comment (akallabeth) [inline libfreerdp/primitives/primitives.c]: should be `primitives_init_gpu_cb`

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Pull request merged**